File: backend/app/schemas/compliance.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.models.compliance_audit import AuditDepth, AuditScope, ControlStatus

# Same vocabulary as AI analysis so the frontend reuses one polling envelope/hook: the DB's
# "running" is surfaced as "in_progress" (see routers/compliance.py _JOB_STATUS_TO_API).
ComplianceAuditStatus = Literal["queued", "in_progress", "completed", "failed"]


# --- catalog (static; drives the wizard) ---


class ControlSummaryOut(BaseModel):
    id: str
    title: str
    reference: str
    code_assessable: bool
    domain: str = "General Controls"
    description: str = ""
    recommendation: str = ""
    manual_reason: str | None = None


class FrameworkOut(BaseModel):
    key: str
    title: str
    scope_note: str
    controls_total: int
    assessed_total: int  # how many of those a code scanner can speak to
    controls: list[ControlSummaryOut] = Field(default_factory=list)


class FrameworkListResponse(BaseModel):
    items: list[FrameworkOut]


# --- audit run ---


class ComplianceAuditCreateRequest(BaseModel):
    """Every field is optional: an empty body means "run the audit this project is
    configured for", which is what the Run Audit button sends.

    The three settings that used to be asked at run time by a three-step wizard --
    frameworks, evidence scope, and whether to pay for the AI narrative -- are now
    configured on the project's Compliance Config tab and resolved from
    workspace_settings_service.effective_compliance_policy. They stay accepted here so a
    CI client or a one-off run can still be explicit, but nothing has to be.
    """

    #: Empty = the configured frameworks (and, if none are configured, every supported one).
    frameworks: list[str] = Field(default_factory=list)
    #: None = the configured evidence scope.
    scope: AuditScope | None = None
    project_repo_ids: list[str] = Field(default_factory=list)  # empty = every repo
    #: None = with_ai_narrative when the workspace has authorised it, deterministic otherwise.
    depth: AuditDepth | None = None
    #: Re-run even when an identical completed audit over the same scans exists. Default False
    #: returns that audit instead, because the evaluator is deterministic — the verdicts would
    #: be byte-identical and the AI narrative would be paid for twice.
    refresh: bool = False


class ControlEvidenceOut(BaseModel):
    fingerprint: str
    scan_id: str
    rule_id: str | None = None
    severity: str | None = None
    file: str
    line: int | None = None
    message: str


class ControlResultOut(BaseModel):
    framework: str
    control_id: str
    control_title: str
    control_reference: str
    status: ControlStatus
    domain: str = "General Controls"
    description: str = ""
    recommendation: str = ""
    rationale: str
    ai_explanation: str | None = None
    ai_remediation: str | None = None
    evidence: list[ControlEvidenceOut] = Field(default_factory=list)
    evidence_total: int = 0
    severity_counts: dict[str, int] = Field(default_factory=dict)


class FrameworkSummaryOut(BaseModel):
    framework: str
    framework_title: str
    scope_note: str
    controls_total: int
    assessed_total: int
    passed: int
    failed: int
    partial: int
    not_applicable: int
    needs_manual_review: int
    #: Percentage of code-assessable controls passed — NOT a compliance percentage.
    compliance_score: int = 0
    #: assessed_total / controls_total, as a percentage: the ceiling on the score above.
    coverage_percent: int = 0


class ComplianceAuditListItem(BaseModel):
    """The audit without its control bodies — for the project's audit history list."""

    id: str
    project_id: str
    frameworks: list[str]
    scope: AuditScope
    depth: AuditDepth
    status: ComplianceAuditStatus
    error_message: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime
    progress_completed: int = 0
    progress_total: int = 0
    findings_total: int = 0
    summaries: list[FrameworkSummaryOut] = Field(default_factory=list)


class ComplianceAuditResponse(ComplianceAuditListItem):
    scan_ids: list[str] = Field(default_factory=list)
    #: Repo coverage of the selected scope. repos_with_scans < repos_in_scope means the audit
    #: saw only part of the project, and the UI says so.
    repos_in_scope: int = 0
    repos_with_scans: int = 0
    newest_scan_at: datetime | None = None
    findings_truncated: bool = False
    ai_note: str | None = None
    #: True when the trigger returned an existing identical audit instead of re-running one.
    #: Not persisted — set by the router on that response only.
    reused: bool = False
    controls: list[ControlResultOut] = Field(default_factory=list)

File: frontend/lib/api/compliance.ts
/**
 * Project-level compliance audits. A distinct pipeline from ai.ts (analysis) and
 * auto-fix.ts (remediation), but the third consumer of the same async-job envelope
 * (AiAnalysisStatus + started_at/progress_*) so it reuses the shared polling helpers.
 *
 * Control status is computed deterministically by the backend from scanner findings.
 * `ai_explanation` / `ai_remediation` are prose only -- they never determine `status`.
 */
import type { AiAnalysisStatus } from "./ai";
import { apiFetch } from "./client";

export type ControlStatus =
  | "pass"
  | "fail"
  | "partial"
  | "not_applicable"
  | "needs_manual_review";

export type AuditScope = "latest" | "history";
export type AuditDepth = "deterministic" | "with_ai_narrative";

export type ControlSummary = {
  id: string;
  title: string;
  reference: string;
  domain: string;
  description: string;
  recommendation: string;
  // False for governance/process controls no code scanner can evidence.
  code_assessable: boolean;
  manual_reason: string | null;
};

export type Framework = {
  key: string;
  title: string;
  scope_note: string;
  controls_total: number;
  assessed_total: number;
  controls: ControlSummary[];
};

export type ControlEvidence = {
  fingerprint: string;
  scan_id: string;
  rule_id: string | null;
  severity: string | null;
  file: string;
  line: number | null;
  message: string;
};

export type ControlResult = {
  framework: string;
  control_id: string;
  control_title: string;
  control_reference: string;
  domain: string;
  description: string;
  recommendation: string;
  status: ControlStatus;
  rationale: string;
  ai_explanation: string | null;
  ai_remediation: string | null;
  evidence: ControlEvidence[];
  // Exact match count; may exceed evidence.length, which is capped by the backend.
  evidence_total: number;
  severity_counts: Record<string, number>;
};

export type FrameworkSummary = {
  framework: string;
  framework_title: string;
  scope_note: string;
  controls_total: number;
  assessed_total: number;
  passed: number;
  failed: number;
  partial: number;
  not_applicable: number;
  needs_manual_review: number;
  // Percentage of *code-assessable* controls that passed. NOT a compliance percentage -- the
  // manual-only controls are not in the denominator. Always render it next to coverage_percent.
  compliance_score: number;
  // assessed_total / controls_total, as a percentage: how much of the framework a code scanner
  // can speak to at all, and so the ceiling on the number above.
  coverage_percent: number;
};

/** The audit without its control bodies -- what the project's history list renders. */
export type ComplianceAuditSummary = {
  id: string;
  project_id: string;
  frameworks: string[];
  scope: AuditScope;
  depth: AuditDepth;
  status: AiAnalysisStatus;
  error_message: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  progress_completed: number;
  progress_total: number;
  findings_total: number;
  summaries: FrameworkSummary[];
};

export type ComplianceAudit = ComplianceAuditSummary & {
  scan_ids: string[];
  // Repo coverage of the selected scope. repos_with_scans < repos_in_scope means the audit only
  // saw part of the project and the result under-represents it.
  repos_in_scope: number;
  repos_with_scans: number;
  newest_scan_at: string | null;
  findings_truncated: boolean;
  ai_note: string | null;
  // True when the backend returned an existing identical audit instead of re-running one.
  reused: boolean;
  controls: ControlResult[];
};

/**
 * Everything is optional. Sending `{}` runs the audit the project is *configured* for —
 * frameworks, evidence scope and AI depth all come from its Compliance Config, resolved
 * server-side so this file never re-derives the workspace/project precedence.
 */
export type RunAuditInput = {
  frameworks?: string[];
  scope?: AuditScope;
  project_repo_ids?: string[];
  depth?: AuditDepth;
  // Force a fresh run even when an identical completed audit over the same scans exists.
  refresh?: boolean;
};

export function listFrameworks() {
  return apiFetch<{ items: Framework[] }>("/compliance/frameworks");
}

export function listProjectAudits(projectId: string, page = 1, pageSize = 20) {
  return apiFetch<{
    items: ComplianceAuditSummary[];
    total: number;
    page: number;
    page_size: number;
  }>(`/projects/${projectId}/compliance-audits?page=${page}&page_size=${pageSize}`);
}

export function getAudit(auditId: string) {
  return apiFetch<ComplianceAudit>(`/compliance/audits/${auditId}`);
}

export function runAudit(projectId: string, input: RunAuditInput = {}) {
  return apiFetch<ComplianceAudit>(`/projects/${projectId}/compliance-audits`, {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export const CONTROL_STATUS_LABEL: Record<ControlStatus, string> = {
  pass: "Pass",
  fail: "Fail",
  partial: "Partial",
  not_applicable: "Not applicable",
  needs_manual_review: "Needs manual review",
};
